### DISCUSSION TITLE
Memory leaks with resumes?

### DISCUSSION THREAD

**Discussion by TheDying0fLight** [effekt-lang/effekt#764] [Q&A]
The following code works fine. But by changing small things I seem to get memory leaks and I dont know why. I would like to know if this is a me issue or Effekt issue does anyone know about this? I provide example memory usage after about 20s of running each code.

### Working:
```
effect Dual(idx: Int): Array[Bool]

def main() = {
  with on[OutOfBounds].panic
  val listsize = 50
  var arr = array(listsize, false)
  try {
    each(0,listsize) { v =>
      arr = do Dual(v)
      println(v)
    }
  } with Dual { idx =>
    arr.set(idx, true)
    resume(arr.copy)
    arr.set(idx, false)
    resume(arr)
  }
}
```
![396066199-427a437c-8d36-49f9-8ea5-f6d0eeabb06b](https://github.com/user-attachments/assets/06de1ba6-732c-4fff-891c-e88f95afdab8)

### Problem1:
If I either change it to this (add a counter to the effect):
```
effect Dual(idx: Int): Array[Bool]

def main() = {
  with on[OutOfBounds].panic
  var countFalse = 0
  val listsize = 50
  var arr = array(listsize, false)
  try {
    each(0,listsize) { v =>
      arr = do Dual(v)
      println(v)
    }
  } with Dual { idx =>
    arr.set(idx, true)
    resume(arr.copy)
    countFalse = countFalse + 1
    arr.set(idx, false)
    resume(arr)
    countFalse = countFalse - 1
  }
}
```
![396066624-1a3fa93c-e1ae-47a3-9a14-73d6089a858e](https://github.com/user-attachments/assets/0e16bb6e-28b1-40c3-9718-03c9ad1a196a)

### Problem2:
Or to this (exchange array with list): 
```
effect Dual(idx: Int): List[Bool]

def main() = {
  with on[OutOfBounds].panic
  var countFalse = 0
  val listsize = 50
  var arr = fill(listsize, false)
  try {
    each(0,listsize) { v =>
      arr = do Dual(v)
      println(v)
    }
  } with Dual { idx =>
    resume(arr.updateAt(idx){_ => true})
    resume(arr.updateAt(idx){_ => false})
  }
}
```
![396066943-fcebf11c-0a8d-4e6f-a3b4-c6cca541c7fa](https://github.com/user-attachments/assets/bd859618-343b-4ae3-a02c-de319a7dce44)

**Comment by marvinborner** [comment]
Thanks for the issue! We often have issues with memory leaks, so I wouldn't be surprised.

The images aren't loading, so I'm not entirely sure what you're referring to (I couldn't observe abnormal memory usage for Problem1 in LLVM, for example). What backend are you using and how are you measuring the memory leaks?

**Comment by TheDying0fLight** [comment]
Exchanged the images with new versions hope they work now. I use the JavaScript backend on Windows 10 and use the Task Manager to track the memory (as the images would show)

**Comment by b-studios** [comment]
In general, we are not the most memory efficient language. However, in your example, you are creating $2^{n + 1} - 2$ many arrays of size $n$ (where n=50). 

So, assuming something like 500 bytes per array, this is roughly **1000 Petabytes**.

JS is garbage collected, so in general we cannot guarantee that the memory will be freed right-away. However, given that the program terminates on your computer says a lot about it not leaking too much memory ;)

It is also not very surprising that the list version allocates a lot more, since there we allocate roughly 50 linked objects.

The following will not make a lot of sense to you, but what we _could_ investigate is whether the "shadow stack" of regions potentially holds onto the arrays for too long preventing GC to do its work. Again, given that the program terminates _at all_ I'd say we are good here.

(disclaimer: I only thought a few minutes about this, so I might be wrong with my analysis)

**Comment by b-studios** [comment]
I have to say, I like this as a benchmark though ;) 

Here (for N=24) is JS (`/usr/bin/time -l`):

```
       11.20 real        10.98 user         0.12 sys
            77856768  maximum resident set size
                   0  average shared memory size
                   0  average unshared data size
                   0  average unshared stack size
                5279  page reclaims
                   1  page faults
                   0  swaps
                   0  block input operations
                   0  block output operations
                   0  messages sent
                   0  messages received
                   0  signals received
                   0  voluntary context switches
               17846  involuntary context switches
        183751342222  instructions retired
         35430678327  cycles elapsed
            60757632  peak memory footprint
```

and llvm:
```
        4.95 real         4.43 user         0.51 sys
          7095566336  maximum resident set size
                   0  average shared memory size
                   0  average unshared data size
                   0  average unshared stack size
              433213  page reclaims
                   4  page faults
                   0  swaps
                   0  block input operations
                   0  block output operations
                   0  messages sent
                   0  messages received
                   0  signals received
                   0  voluntary context switches
                  37  involuntary context switches
         70141067796  instructions retired
         15580367548  cycles elapsed
          7098611584  peak memory footprint
```

Here LLVM uses 5,6 GB peak memory, while JS is around 58 MB! (cc @phischu). Nevertheless, it is still twice as fast.

**Comment by TheDying0fLight** [comment]
> In general, we are not the most memory efficient language. However, in your example, you are creating 2 n + 1 − 2 many arrays of size n (where n=50).
> 
> So, assuming something like 500 bytes per array, this is roughly **1000 Petabytes**.
> 
> JS is garbage collected, so in general we cannot guarantee that the memory will be freed right-away. However, given that the program terminates on your computer says a lot about it not leaking too much memory ;)
> 
> It is also not very surprising that the list version allocates a lot more, since there we allocate roughly 50 linked objects.

Sorry I was not clear about this. The programm did not terminate for me, but that is not the important part. What I wanted to demonstrate is that adding a counter or changing an array to a list will cause a memory leak. While the example without them rarely uses more than 50MB of RAM the other ones quickly run out of memory when the `println` is removed.

I tested a bit with the counting. When running the following code we get an ever increasing memory usage (task manager screenshots after running for 20s or Out of memory, what happens first):
![image](https://github.com/user-attachments/assets/7dbfec57-901e-4a36-a2bf-3877de12db4e)

```
effect Dual(idx: Int): Array[Bool]

def main() = {
  with on[OutOfBounds].panic
  var count = 0
  val listsize = 50
  var arr = array(listsize, false)
  try {
    each(0,listsize) { v =>
      arr = do Dual(v)
    }
  } with Dual { idx =>
    arr.set(idx, true)
    resume(arr.copy)
    arr.set(idx, false)
    resume(arr)
    count = count + 1
  }
}
```
If the `count` is moved outside (directly before the effect), which does results in the same count, the memory usage increases slowly or not at all (Only code for `try` `with` as the rest stays the same):
![image](https://github.com/user-attachments/assets/9b8379f1-671f-45c8-aae4-5a70d75b4b38)

```
try {
each(0,listsize) { v =>
  count = count + 1
  arr = do Dual(v)
 }
} with Dual { idx =>
  arr.set(idx, true)
  resume(arr.copy)
  arr.set(idx, false)
  resume(arr)
}
```
It seems the counting at the end of the effect has some weird behaviour. When the `count` is increased before the end of the effect this issue also does not occur:
![image](https://github.com/user-attachments/assets/4b22e9d2-9fe2-4740-9a1e-ec5b2e036031)
```
try {
  each(0,listsize) { v =>
    arr = do Dual(v)
  }
} with Dual { idx =>
  arr.set(idx, true)
  resume(arr.copy)
  arr.set(idx, false)
  count = count + 1
  resume(arr)
}
```
> The following will not make a lot of sense to you, but what we _could_ investigate is whether the "shadow stack" of regions potentially holds onto the arrays for too long preventing GC to do its work.

What you described seems plausible (but for lists and not for arrays since arrays behave as expected). It shouldn't use much more memory than the arrays solution as it creates a similar amount of arrays (also tried copying the array in both resumes with similar results).

**Comment by phischu** [comment]
Thank you for minifying these examples. The expected result if you would print `count` at the end is `2^50`, right?

```
  } with Dual { idx =>
    arr.set(idx, true)
    resume(arr.copy)
    arr.set(idx, false)
    resume(arr)
    count = count + 1
  }
  ```
  
In a handler like this, the increments will only be done after the very last `resume` has finished. Operationally, we remember that we will have to increment the count as a stack frame. This requires a stack of size `2^50`.

**Comment by TheDying0fLight** [comment]
Wouldn't the following code then have the same issue if that is the problem? I reduced the iterations to `26` then code without memory issues finishes on my PC and the other fails due to OOM.
```
effect Dual(): Unit

def main() = {
  var count = 0
  try {
    each(0,26) { _ =>
      do Dual()
    }
  } with Dual {
    resume(())
    resume(())
    count = count + 1
  }
}
```
I removed the first `set` as it does not have an effect on the issue. I think I pinned the issue down to these two lines, when both lines are there the memory issues exist otherwise its fine.
```
effect Dual(idx: Int): Array[Bool]

def main() = {
  with on[OutOfBounds].panic
  var count = 0
  val listsize = 26
  var arr = array(listsize, false)
  try {
    each(0,listsize) { v =>
      arr = do Dual(v)
    }
  } with Dual { idx =>
    resume(arr.copy)
    arr.set(idx, false) // if this is removed there is no issue
    resume(arr)
    count = count + 1 // same for this (only one has to be removed)
  }
}
```
